### ISSUE TITLE
Different learned strategies from the same experiment

### ISSUE THREAD

**Issue by Aloriosa** [alex-petrenko/sample-factory#322]
Hello!
Thank you for your amazing work!

I am running the experiments on grid-world MAPF environments with synchronous experience collection, and different runs with the same configuration on the same hardware result in significantly different agent strategies. I use the same random seed across runs and make no changes to code or train configurations. I was looking through the training logs and found that only the runs that resulted in the desired strategy (i.e. strategy satisfies the cumulative reward maximization rule) have exactly matching policy lags, while the rest of the runs have various non-matching patterns of policy lag. 
I have tried running the same experiments multiple times, but was not able to reproduce that successive training behavior.
I have also tried to do A2C with `--async_rl=False` and `--num_batches_per_epoch=1` and `--num_epochs=1` as it was stated in the docs, but this setting did not succeed in my case.

Can you give some recommendations on how to reproduce the desired policy lag or at least stabilize if from run to run with all the other settings fixed? 
Is it possible to control the policy lag somehow during training? Maybe manually bounding the available CPU resources or memory usage?

**Comment by Aloriosa** [comment]
@alex-petrenko Hello!
Thank you for your quick response!

Yes, I use --max_policy_lag=1 along with --async_rl=False and --num_batches_per_epoch=16 and --num_epochs=1. 
When training with --async_rl=False and --num_batches_per_epoch=1 and --num_epochs=1, the policy lag is equal to zero:

<img width="477" height="399" alt="Image" src="https://github.com/user-attachments/assets/9c61595e-c14c-4109-84ec-e4e210901a86" />

I have two issues with training:
1) Running the same config with a fixed random seed on the same hardware several times results in significantly different agents' behavior during the episode.  

2) The training reward initially grows and then stucks around some low value.

Could you please give some recommendations on how to stabilize training and what training parameters can I explore to make training reward progress?


Regarding the training parameters.
I am using the custom CNN encoder and the transformer core model with the rollout+1 input sequence length (meaning the current obs and the `rollout` previous observations).
My training config is the following:
```
{
  "help": false,
  "algo": "APPO",
  "env": "myenv",
  "experiment": "myexp",
  "train_dir": "experiments/train_dir",
  "restart_behavior": "resume",
  "device": "gpu",
  "seed": 3,
  "num_policies": 1,
  "async_rl": false,
  "serial_mode": false,
  "batched_sampling": false,
  "num_batches_to_accumulate": 1,
  "worker_num_splits": 1,
  "policy_workers_per_policy": 1,
  "max_policy_lag": 1,
  "num_workers": 4,
  "num_envs_per_worker": 4,
  "batch_size": 16384,
  "num_batches_per_epoch": 16,
  "num_epochs": 1,
  "rollout": 8,
  "recurrence": 1,
  "shuffle_minibatches": false,
  "gamma": 0.9716,
  "reward_scale": 1.0,
  "reward_clip": 1000.0,
  "value_bootstrap": true,
  "normalize_returns": false,
  "exploration_loss_coeff": 0.0156,
  "value_loss_coeff": 0.5,
  "kl_loss_coeff": 0.0,
  "exploration_loss": "entropy",
  "gae_lambda": 0.95,
  "ppo_clip_ratio": 0.2,
  "ppo_clip_value": 1.0,
  "with_vtrace": false,
  "vtrace_rho": 1.0,
  "vtrace_c": 1.0,
  "optimizer": "adam",
  "adam_eps": 1e-06,
  "adam_beta1": 0.9,
  "adam_beta2": 0.999,
  "max_grad_norm": 4.0,
  "learning_rate": 0.00013,
  "lr_schedule": "kl_adaptive_minibatch",
  "lr_schedule_kl_threshold": 0.008,
  "lr_adaptive_min": 1e-06,
  "lr_adaptive_max": 0.01,
  "obs_subtract_mean": 0.0,
  "obs_scale": 1.0,
  "normalize_input": false,
  "normalize_input_keys": null,
  "decorrelate_experience_max_seconds": 0,
  "decorrelate_envs_on_one_worker": true,
  "actor_worker_gpus": [],
  "set_workers_cpu_affinity": true,
  "force_envs_single_thread": true,
  "default_niceness": 0,
  "log_to_file": true,
  "experiment_summaries_interval": 10,
  "flush_summaries_interval": 30,
  "stats_avg": 10,
  "summaries_use_frameskip": true,
  "heartbeat_interval": 20,
  "heartbeat_reporting_interval": 180,
  "train_for_env_steps": 30000000,
  "train_for_seconds": 10000000000,
  "save_every_sec": 1800,
  "keep_checkpoints": 1000000,
  "load_checkpoint_kind": "latest",
  "save_milestones_sec": -1,
  "benchmark": false,
  "use_rnn": false,
  "rnn_size": 16,
  "rnn_type": "gru",
  "rnn_num_layers": 1,
  "decoder_mlp_layers": [],
  "nonlinearity": "elu",
  "policy_initialization": "orthogonal",
  "policy_init_gain": 1.0,
  "actor_critic_share_weights": true,
  "adaptive_stddev": true,
  "continuous_tanh_scale": 0.0,
  "initial_stddev": 1.0,
  "use_env_info_cache": false,
  "env_gpu_actions": false,
  "env_gpu_observations": true,
  "env_frameskip": 1,
  "env_framestack": 1,
  "pixel_format": "CHW",
  "use_record_episode_statistics": false,
  "with_wandb": false,
  "wandb_user": null,
  "wandb_project": null,
  "wandb_group": null,
  "wandb_job_type": "SF",
  "wandb_tags": [],
  "with_pbt": false,
  "pbt_mix_policies_in_one_env": true,
  "pbt_period_env_steps": 5000000,
  "pbt_start_mutation": 20000000,
  "pbt_replace_fraction": 0.3,
  "pbt_mutation_rate": 0.15,
  "pbt_replace_reward_gap": 0.1,
  "pbt_replace_reward_gap_absolute": 1e-06,
  "pbt_optimize_gamma": false,
  "pbt_target_objective": "true_objective",
  "pbt_perturb_min": 1.1,
  "pbt_perturb_max": 1.5,
  "worker_index": null,
    "vector_index": null,
    "env_id": null,
    "every_step_metrics": false
  }
```

**Comment by alex-petrenko** [comment]
> Yes, I use --max_policy_lag=1 along with --async_rl=False and --num_batches_per_epoch=16 and --num_epochs=1.
Probably not a good idea. With 16 batches per epoch the algorithm does 16 SGD steps, and 15 of them (all except first) are therefore lagged. You will just be discarding 15/16th of your experience, very inefficient. You should only use max_policy_lag=1 with num_batches_per_epoch=1. 

> When training with --async_rl=False and --num_batches_per_epoch=1 and --num_epochs=1, the policy lag is equal to zero:
Okay, thank you for confirming. This makes sense.


> Running the same config with a fixed random seed on the same hardware several times results in significantly different agents' behavior during the episode.

This is totally expected. Achieving total reproducibility down to numerical noise is very challenging in RL. SF2 makes no attempt to do so. Typically none of the underlying systems (pytorch/CUDA) provide full numerical reproducibility because of how floating point operations are implemented (the compiler is free to reorder some operations which affects least significant digits when doing the same computation).

You're free to experiment with https://docs.pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms and similar features, but I'd advise you not to pursue this.

At the very best, if you succeed, this gives false sense of stability. You're risking to overfit to specific seeds and numerical noise. Any tiny change in your system (e.g. a random library update) will slightly affect this numerical noise yielding different results.
My advice is to run your experiments deliberately with random seeds, always, to have a sense of not only overall performance but also variance.

Instability of results like yours is very common in RL. Typically it's a feature of your environment/reward function rather than the RL algorithm. Although some algorithms can perform better than others due to different exploration trajectory etc.

> The training reward initially grows and then stucks around some low value.

This is a very common problem, we call it "stuck in a local optimum". RL algorithms like PPO are stochastic local optimizers. They climb local slopes of loss functions. There's no guarantee a global optimum can be found.

Consider how you can change your MDP/reward to better guide the agent out of suboptimal ravines. There are powerful exploration methods like PBT that can still help when everything else fails. https://arxiv.org/abs/2305.12127
